### PR TITLE
fix: fix encoding/matching issues with special chars

### PR DESCRIPTION
**Pull request by brophdawg11**
**tl;dr;**
Replaces #9458 with a fairly comprehensive test suite to ensure that we can handle paths with special characters, match against them, and received properly encoded values from `useLocation` that match what `window.location` would report.

**Background**
In 6.3 and prior we were always in a _reactive_ mode to `window.history` so for example, we would `history.pushState` and then call the listeners with the value from `window.location`.  This flow meant that we inherited the built-in URL encoding performed by `window.location`.  Therefore, if you clicked `<Link to="/✅">` you got `{ pathname: "/%E2%9C%85", ... }` from `useLocation()`.

In 6.4, we go through the new `@remix-run/router` first and do all of our data fetching and `updateState` _before_ `history.pushState` - so we weren't getting the encoding.  So at the moment, if you click `<Link to="/✅">` you incorrectly get `{ pathname: "/✅", ... }` from `useLocation()`.

We also have historically had trouble with defining route paths with special characters since we were matching encoded paths from `window.location` to unencoded paths in our `<Route>` objects.  

This PR fixes these 3 main things surrounding special characters:
* `history` should always send through the `window.location` to listeners to ensure full backwards compatibility with 6.3
* When routing in a data router via `router.navigate` we manually encode the incoming location using `new URL()` so we get the same encoding behavior as `window.location`.  This ensures that useLocation in .4 matches what it would have returned in a 6.3 app
* Since we are now consistently operating on incoming encoded paths, but we want to preserve the ability to match routes defined with special characters we gained from #9300, we now decode the incoming pathname prior to matching it against our `<Route path>` values - allowing us to do `<Route path="✅" />`

### CODE REVIEW

File: packages/react-router-dom/__tests__/navigate-encode-params-test.tsx [hunk 1]
```diff
@@ -88,7 +89,8 @@ describe("navigate with params", () => {
       let pathname = window.location.pathname.replace(/%20/g, "+");
       expect(pathname).toEqual("/blog/react+router");
 
-      expect(node.innerHTML).toMatch(/react router/);
+      // Note decodeURIComponent doesn't decode +
+      expect(node.innerHTML).toMatch(/react\+router/);
```
Comment (brophdawg11) [inline packages/react-router-dom/__tests__/navigate-encode-params-test.tsx:93]: This was a false positive test previously - and was seemingly only because we weren't resetting history state between tests.  `decodeURIComponent` doesn't decode the `+` sign

File: packages/router/history.ts [hunk 2]
```diff
@@ -570,7 +570,7 @@ function getUrlBasedHistory(
     }
 
     if (v5Compat && listener) {
-      listener({ action, location });
+      listener({ action, location: history.location });
```
Comment (brophdawg11) [inline packages/router/history.ts:573]: History should pass the encoded location (from `window.location` after `pushState`) and not the incoming unencoded location

File: packages/router/router.ts [hunk 3]
```diff
@@ -750,6 +750,20 @@ export function createRouter(init: RouterInit): Router {
...
+    // remains the same as POP and non-data-router usages.  new URL() does all
+    // the same encoding we'd get from a history.pushState/window.location read
+    // without having to touch history
+    let url = createURL(createPath(location));
+    location = {
+      ...location,
+      pathname: url.pathname,
+      search: url.search,
+      hash: url.hash,
+    };
```
Comment (brophdawg11) [inline packages/router/router.ts:765]: Encode all incoming `router.navigate()` locations the same way `window.location` would

File: packages/router/utils.ts [hunk 4]
```diff
@@ -329,7 +329,13 @@ export function matchRoutes<
 
   let matches = null;
   for (let i = 0; matches == null && i < branches.length; ++i) {
-    matches = matchRouteBranch<string, RouteObjectType>(branches[i], pathname);
+    matches = matchRouteBranch<string, RouteObjectType>(
+      branches[i],
+      // incoming pathnames are always encoded from either window.location or
+      // from route.navigate, but we want to match against the unencoded paths
+      // in the route definitions
+      safelyDecodeURI(pathname)
```
Comment (brophdawg11) [inline packages/router/utils.ts:337]: Decode incoming pathnames for matching against our route definitions

File: packages/react-router-dom/__tests__/special-characters-test.tsx [hunk 5]
```diff
@@ -0,0 +1,712 @@
...
+ *   return {
+ *       char,
+ *       pathChar: new URL('/' + char, window.location.origin).pathname.replace(/^\//, ''),
+ *       searchChar: new URL('/?=' + char, window.location.origin).search.replace(/^\?=/, ''),
+ *       hashChar: new URL('/#' + char, window.location.origin).hash.replace(/^#/, ''),
+ *   };
+ * }
+ */
+
+let specialChars = [
```
Comment (brophdawg11) [inline packages/react-router-dom/__tests__/special-characters-test.tsx:41]: Large set of special character "definitions" of the character, and what we expect it to show up as in `location.pathname`, `location.search`, and `location.hash` since it varies according to the spec: https://url.spec.whatwg.org/#c0-control-percent-encode-set

File: packages/react-router-dom/__tests__/special-characters-test.tsx [hunk 6]
```diff
@@ -0,0 +1,712 @@
...
+              element={<Comp heading="Parent Nested Splat Route" />}
+            />
+          </Route>
+          <Route
+            path="/reset"
+            element={<Link to={navigatePath}>Link to path</Link>}
+          />
+          <Route path="/*" element={<Comp heading="Root Splat Route" />} />
+        </>
+      );
```
Comment (brophdawg11) [inline packages/react-router-dom/__tests__/special-characters-test.tsx:221]: Test that we can support matching special characters when used in dynamic/splat params in root and nested scenarios

File: packages/react-router-dom/__tests__/special-characters-test.tsx [hunk 7]
```diff
@@ -0,0 +1,712 @@
...
+
+      // Render BrowserRouter at the initialized location and confirm we get
```
Comment (brophdawg11) [inline packages/react-router-dom/__tests__/special-characters-test.tsx:223]: The section here and below basically checks 4 things:
* We match routes properly on initial `BrowserRouter` load (reading from `window.location`)
* We match routes properly on client-side `BrowserRouter` navigations (through `history.push`)
* We match routes properly on initial `createBrowserRouter`/`RouterProvider` load (reading from `window.location`)
* We match route properly on client-side `createBrowserRouter`/`RouterProvider` navigations (through `router.navigate`)

File: packages/react-router-dom/__tests__/special-characters-test.tsx [hunk 8]
```diff
@@ -0,0 +1,712 @@
...
+          <Route path="/nested">
+            <Route path={char} element={<StaticNested />} />
+          </Route>
+          <Route path="/:param">
+            <Route path={char} element={<ParamNested />} />
+          </Route>
+          <Route path="/reset" element={<Link to={path}>Link to path</Link>} />
+          <Route path="*" element={<h1>Not Found</h1>} />
+        </>
+      );
```
Comment (brophdawg11) [inline packages/react-router-dom/__tests__/special-characters-test.tsx:582]: Test that we can define and match against route paths containing special characters

File: packages/router/router.ts (hunk 3 above)
Comment (brophdawg11) [reply packages/router/router.ts:765]: Note to self - move this to a `history.encodeLocation` method since we don't need to encode for memory histories

File: packages/router/utils.ts [hunk 9]
```diff
@@ -702,6 +708,21 @@ function compilePath(
   return [matcher, paramNames];
 }
 
+function safelyDecodeURI(value: string) {
+  try {
+    return decodeURI(value);
```
Comment (mjackson) [inline packages/router/utils.ts:713]: How sure are you that `decodeURI` is what we want here? It's been a while since I dug into the URL encoding stuff, but last time I did I remember reading something about "you almost always want decodeURIComponent, not decodeURI". Does that sound familiar?

Comment (brophdawg11) [reply packages/router/utils.ts:713]: I'll do some digging but I think that would work.  The choice of `decodeURI` instead of `decodeURIComponent` was to align with when I was initially encoding via `encodeURI` instead of `encodeURIComponent` (since `encodeURIComponent` is a bit too aggressive).  

But now that encoding is done via `new URL` it may be fine to switch to `decodeURIComponent`

Screen Shot 2022-10-20 at 3 37 33 PM

Comment (brophdawg11) [reply packages/router/utils.ts:713]: I think want to keep `decodeURI` here - since `decodeURIComponent` is also a bit too aggressive and _also_ decodes properly-encoded URL chars such as `/` and `?` whereas decodeURI will keep them encoded.  See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI#description for an explanation of the treatment of special meaning URL characters.

We actually have a test around this too in `params-decode-test` for when a `/` is used in a URL like `/content/%2F`.  Here's a quick example of how `decodeURIComponent` goes too far in this case and creates a path with unintended slash separators and query strings:

Screen Shot 2022-10-20 at 4 10 30 PM

Comment (mjackson) [reply packages/router/utils.ts:713]: Sounds good. Thanks for digging 👍

**Review by mjackson — APPROVED** [review]
